### PR TITLE
add sentry reporting to connector publish pipelines

### LINKED ISSUE #28689: Dagger Pipeline Errors Should be tracked in Sentry
## Solution Criteria
1. We should log all Dagger pipeline errors that occur outside the steps to sentry
2. We should not be tracking or alerting on errors that we deem specific to a connector 
3. Should be able to ensure we only get alerts on specific branches and maybe specific errors/codepaths

### PR DESCRIPTION
**Pull request by pedroslopez**
## What

closes #28689

Adds sentry instrumentation for the connector publish pipeline.
The same can be easily added to test and other pipeline executions by passing the SENTRY_DSN env var, but since I haven't tested with tests yet and want to start to get this out there, left scoped to publish for now.

## How

- Install sentry python sdk
- Set up some util methods to set sentry tags and context for pipeline command and step execution

You can check out a (real) error reported by this here: https://airbytehq.sentry.io/issues/4342827390/?project=4505596642983936&query=&referrer=issue-stream&stream_index=3

### What you see in sentry

image

By default sentry adds the release with the current commit and other helpful info. I've added these custom tags:
- `connector` 
- `pipeline_command` 
- `pipeline_step`

You'll also see a few contexts populated:

image
image
image
image

### CODE REVIEW

File: airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py
Comment (pedroslopez) [inline airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py:1]: Wanted to keep the sentry stuff contained. Also makes it easier to set up the scopes so we're sure we report step-related info if we have a failure on that step.

File: .github/workflows/publish_connectors.yml [hunk 1]
```diff
@@ -53,6 +54,7 @@ jobs:
           gcs_credentials: ${{ secrets.METADATA_SERVICE_PROD_GCS_CREDENTIALS }}
           github_token: ${{ secrets.GH_PAT_MAINTENANCE_OCTAVIA }}
           metadata_service_gcs_credentials: ${{ secrets.METADATA_SERVICE_PROD_GCS_CREDENTIALS }}
+          sentry_dsn: "https://f2e9aeb6b24b4024b09679201c7f563e@o1009025.ingest.sentry.io/4505596642983936"
```
Comment (evantahler) [inline .github/workflows/publish_connectors.yml:57]: not a secret?

Comment (pedroslopez) [reply .github/workflows/publish_connectors.yml:57]: It's not, but if you have admin access and can help setting it up as one we can get rid of this duplication 😄 @evantahler

Comment (evantahler) [reply .github/workflows/publish_connectors.yml:57]: Done! `SENTRY_CONNECTOR_DAGSTER_DSN` is what you want now

Comment (alafanechere) [reply .github/workflows/publish_connectors.yml:57]: It should be `SENTRY_AIRBYTE_CI_DSN` not `DAGSTER` 😄 will change as I've admin access.

Comment (alafanechere) [reply .github/workflows/publish_connectors.yml]: @pedroslopez I pushed the change.

File: airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                'retry_count': self.retry_count,
+            })
+
+            if hasattr(self.context, "connector"):
+                connector: Connector = self.context.connector
+                scope.set_tag("connector", connector.technical_name)
+                scope.set_context("Connector", {
+                    'name': connector.name,
+                    'technical_name': connector.technical_name,
+                    'language': connector.language,
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py:23]: Shall we add the version + release stage?
```suggestion
                    'language': connector.language,
                    'version': connector.version
                    'release_stage': connector.release_stage
```

File: airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+def with_command_context(func):
+    def wrapper(self, ctx, *args, **kwargs):
+        with sentry_sdk.configure_scope() as scope:
+            scope.set_tag("pipeline_command", self.name)
+            scope.set_context("Pipeline Command", {
+                'name': self.name,
+                'params': self.params,
+            })
+            scope.set_context("Click Context", ctx.obj)
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py:38]: The ctx.obj has secrets like SA account, GitHub PAT etc., will they be obfuscated by Sentry?

File: .github/workflows/publish_connectors.yml [hunk 4]
```diff
@@ -37,6 +37,7 @@ jobs:
...
```
Comment (alafanechere) [inline .github/workflows/publish_connectors.yml]: Setting this in this workflow is what make the reporting only enable for publish right?

**Review by alafanechere — APPROVED** [review]
Neat! Just to be clear, what will be reported are unhandled errors - dagger exec errors are handled by steps.

File: .github/workflows/publish_connectors.yml (hunk 4 above)
Comment (pedroslopez) [reply .github/workflows/publish_connectors.yml]: Yep!

File: airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py (hunk 3 above)
Comment (pedroslopez) [reply airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py:40]: Yep! 
image

File: airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+def with_step_context(func):
+    def wrapper(self, *args, **kwargs):
+        with sentry_sdk.configure_scope() as scope:
+            step_name = self.__class__.__name__
+            scope.set_tag("pipeline_step", step_name)
+            scope.set_context("Pipeline Step", {
+                'name': step_name,
+                'step_title': self.title,
+                'max_retries': self.max_retries,
+                'retry_count': self.retry_count,
```
[the remark this answers is not in the document]
Comment (pedroslopez) [reply airbyte-ci/connectors/pipelines/pipelines/sentry_utils.py:30]: added in [5615658](https://github.com/airbytehq/airbyte/pull/28748/commits/5615658ada4d8c148476bf940a70b2642a46a307)

**Pull request merged**